### DISCUSSION TITLE
Virtual IP traffic not working without explicit SNAT rule

### DISCUSSION THREAD

**Discussion by p-try** [strongswan/strongswan#281] [Q&A]
Hello guys,
I have a really weird problem that I would like to discuss here before filing a bug because I might have missed something.

**Situation:**
We have a server (dedicated, not virtual) that is hosted in a data center outside of our premises and that we connect to our corporate network via VPN. The external IP address of our server is **176.9.xxx.xxx** (anonymized in the attached logs) and it is running **Ubuntu 20.04**. Our corporate router is reachable at vpn.corporation.de (also anonymized) and the internal corporate network is **192.168.29.0/24** (not anonymized). Our router is a LANCOM 1781VA which has IPsec and IKEv2 built-in. The server is assigned a virtual IP address from our internal network via IKE, which is **192.168.29.230**.

**Working:**
The IPsec tunnel is connecting without any problems and the server is receiving the virtual IP address. I can verify this in both the charon log and on the router. I can ping the server using its virtual IP address from within the corporate network and I can even SSH into the server using this address. No problems so far, so the server has obviously received the virtual address and is responding to it properly.

**Now here comes the funny part:**
It doesn't work the other way round. I can't connect to any services in the corporate network when I initiate the connection from the external server, neither ping, nor DNS nor TCP services. I have double-checked our firewall to make sure it doesn't block the traffic. It doesn't.
It will work instantly, however, once I set up a SNAT rule on the external server to mask the outgoing traffic to the VPN network behind the virtual IP address:

`iptables -t nat -A POSTROUTING --destination 192.168.29.0/24 -j SNAT --to-source 192.168.29.230`

And that's what I don't understand. The server already knows the virtual IP address and is responding to it correctly when I connect to it from the corporate network, so why would it need this SNAT rule for the other way? It doesn't say anything about that in the strongswan documentation and I haven't needed this before on other servers.

Of course I could just live with it and keep the SNAT rule but this has given me a real headache so maybe somebody can help me get to the bottom of this.

Thank you very much in advance!

[logs_and_config.zip](https://github.com/strongswan/strongswan/files/6195525/logs_and_config.zip)

**Comment by Thermi** [comment]
The route that should do that doesn't exist in table 220. That can happen if you disable the route installation. Double check if you did that.

**Reply by p-try** [reply]
Thank you very much for your comment and pointing out the necessity of the route, this may help me narrow down the problem. I have explicitly set `install_routes = yes` in charon.conf:
```
root@balblair ~ # grep -r install_routes /etc/strongswan.*
/etc/strongswan.d/charon.conf:    install_routes = yes
```
but `ip route show table 220` is still empty.

I have set up the same thing on another server also running Ubuntu 20.04 and the problem doesn't appear there.

But I have noticed something now. The server is hosted in a data center that requires all traffic to go through an on-link gateway rather than having a subnet and a gateway within that, which is why the data center operator sets up the following in netplan:

```
      routes:
        - on-link: true
          to: 0.0.0.0/0
          via: 176.9.47.1
```

The working server does not have that, it is part of a subnet. Could this route be the reason why strongswan does not set up the route? I could probably easily solve this with an updown script adding the route manually.

**Reply by Thermi** [reply]
I see. There is a route in the main table, but there's no route to the next hop, except that it uses the device enp35s0. Offhand I do not know which part causes the problem. @tobiasbrunner your thoughts on this?
Right now, you can try to add a link route to the next hop using the enp35s0 device.
E.g. `ip route add 176.9.47.1 dev enp35s0`

**Reply by Thermi** [reply]
Did it fix the issue?

**Reply by p-try** [reply]
Sorry for the late reply. Yes, indeed, after adding the route for the next hop manually, Strongswan installs the route correctly and VPN traffic is working without the SNAT rule. Is this something that Strongswan should account for by itself or do you think this kind of network configuration is rather unusual and faulty? If it is of any interest, our datacenter operator is Hetzner (hetzner.com).

**Reply by Thermi** [reply]
It's a rather unususal configuration and is not supported by the code yet. It could be solved though.
@tobiasbrunner

**Reply by tobiasbrunner** [reply]
There are several log messages missing e.g.  `getting a local address in traffic selector ...` or `installing route: ... via ... src ... dev ...` when the outbound policy is installed (these messages have the same log level). They should be seen even without next hop etc. Maybe route installation was actually disabled initially. Could be interesting to see the log messages without that manual route installed.

Anyway, when determining the next hop for the route to be installed, we basically check for a route to the peer for which we can determine a source address (the next hop is the one in that route or there is a fallback to the peer's address). The default route here does not explicitly specify a source IP, and it might not specify an interface or a next hop directly either (if there was an interface, we should find the source IP via that and accept the route). The information listed there might be from an `RTM_MULTIPATH` attribute (where the `onlink` attribute seems to come from). We currently don't parse these, so the route might just get ignored. Why the additional route would change that I don't know.

It might be possible to use the kernel's route lookup (by setting `charon.plugins.kernel-netlink.fwmark` to an inverted mark) instead of installing a helper route.
